channels_redis/serializers.py

```diff
@@ -0,0 +1,140 @@
+import abc
+import json
+import random
+
+
+class SerializerDoesNotExist(KeyError):
+    """The requested serializer was not found."""
+
+
+class BaseMessageSerializer(abc.ABC):
+    def __init__(
+        self,
+        symmetric_encryption_keys=None,
+        random_prefix_length=0,
+        expiry=None,
+    ):
+        self.random_prefix_length = random_prefix_length
+        self.expiry = expiry
+        # Set up any encryption objects
+        self._setup_encryption(symmetric_encryption_keys)
+
+    def _setup_encryption(self, symmetric_encryption_keys):
+        # See if we can do encryption if they asked
+        if symmetric_encryption_keys:
+            if isinstance(symmetric_encryption_keys, (str, bytes)):
+                raise ValueError(
+                    "symmetric_encryption_keys must be a list of possible keys"
+                )
+            try:
+                from cryptography.fernet import MultiFernet
+            except ImportError:
+                raise ValueError(
+                    "Cannot run with encryption without 'cryptography' installed."
+                )
+            sub_fernets = [self.make_fernet(key) for key in symmetric_encryption_keys]
+            self.crypter = MultiFernet(sub_fernets)
+        else:
+            self.crypter = None
+
+    @abc.abstractmethod
+    def dumps(self, message):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def loads(self, message):
+        raise NotImplementedError
+
+    def serialize(self, message):
+        """
+        Serializes message to a byte string.
+        """
+        message = self.dumps(message)
+        # ensure message is bytes
+        if isinstance(message, str):
+            message = message.encode("utf-8")
```

Comment:
I'm not sure I like this specific part of the design. IMO the serializer implementation should handle *all* serialization, including the string encoding settings (`utf-8` or otherwise)

Deferring all serialization to the `dumps` function means that this is symmetric with how `deserialize` calls `load`

(this raises a separate concern that `dumps` and `loads` are not named correctly, perhaps `as_bytes` and `from_bytes` instead?)

Reply:
Seems reasonable, I used `utf-8` as it is also used in [`group_send`](https://github.com/django/channels_redis/blob/13cef4502a95fbb7f41bb97eaa6f4d02c1f7c91c/channels_redis/core.py#L533) and in [`_group_key`](https://github.com/django/channels_redis/blob/13cef4502a95fbb7f41bb97eaa6f4d02c1f7c91c/channels_redis/core.py#L651).

Thus I belive that if we provide support for encoding for messages we may also extend that to those methods to keep consistency (encoding also relates to #222).